text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,648 @@
+>  the function signatures of caller and callee are not compatible. Guaranteed
+>  tail calls enable a class of algorithms that would otherwise use an
+>  arbitrary amount of stack space.
+>
+> If a ``return`` statement is marked ``musttail``, this indicates that the
+>  compiler must generate a tail call for the program to be correct, even when
+>  optimizations are disabled. This guarantees that the call will not cause
+>  unbounded stack growth if it is part of a recursive cycle in the call graph.
+>
+> If the callee is a virtual function that is implemented by a thunk, there is
+>  no guarantee in general that the thunk tail-calls the implementation of the
+>  virtual function, so such a call in a recursive cycle can still result in
+>  unbounded stack growth.
+>
+> ``clang::musttail`` can only be applied to a ``return`` statement whose value
+> is the result of a function call (even functions returning void must use
+> ``return``, although no value is returned). The target function must have the
+> same number of arguments as the caller. The types of the return value and all
+> arguments must be similar according to C++ rules (differing only in cv
+> qualifiers or array size), including the implicit "this" argument, if any.
+> Any variables in scope, including all arguments to the function and the
+> return value must be trivially destructible. The calling convention of the
+> caller and callee must match, and they must not be variadic functions or have
+> old style K&R C function declarations.
+
+There is also a [proposal](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2920.pdf) for the [C Standard](https://www.open-std.org/JTC1/SC22/WG14/) outlining some limitations for Clang.
+> Clang requires the argument types, argument number, and return type to be the same between the
+> caller and the callee, as well as out-of-scope considerations such as C++ features and the calling
+> convention. Implementor experience with Clang shows that the ABI of the caller and callee must be
+> identical for the feature to work; otherwise, replacement may be impossible for some targets and
+> conventions (replacing a differing argument list is non-trivial on some platforms).
+
+
+## GCC
+GCC does not support a feature equivalent to Clang's `musttail`, there also does not seem to be a push to implement it ([pipermail](https://gcc.gnu.org/pipermail/gcc/2021-April/235882.html)) (as of 2021). However, there also exists an experimental [plugin](https://github.com/pietro/gcc-musttail-plugin) for GCC last updated in 2021.
+
+
+## Zig
+Zig provides separate syntax to allow more flexibility than normal function calls. There are options for async calls, inlining, compile-time evaluation of the called function, or specifying TCE on the call.
+([source](https://ziglang.org/documentation/master/#call))
+```zig
+const expect = @import("std").testing.expect;
+
+test "noinline function call" {
+    try expect(@call(.auto, add, .{3, 9}) == 12);
```

Comment:
Would be nicer to show an example with tail calls directly. The Zig documentation doesn't seem to provide one, but maybe someone knowledgeable with it can write one.

Reply:
Yes, it would be nice to have one. I didn't find an example in the documentation either. Also community sentiment is missing, as the TODO below indicates.

Reply:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/51ecc12222b02663da5a581875be8e8b87025959

At least a working example.

Reply:
Resolving as we now have an example, please reopen if you disagree.